runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -303,14 +309,34 @@ public void processElements(Iterable<WindowedValue<InputT>> values) throws Excep
       windowsToConsider.addAll(processElement(windowToMergeResult, value));
     }
 
-    // Trigger output from any window for which the trigger is ready
-    for (W mergedWindow : windowsToConsider) {
-      ReduceFn<K, InputT, OutputT, W>.Context directContext =
-          contextFactory.base(mergedWindow, StateStyle.DIRECT);
-      ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-          contextFactory.base(mergedWindow, StateStyle.RENAMED);
-      triggerRunner.prefetchShouldFire(mergedWindow, directContext.state());
-      emitIfAppropriate(directContext, renamedContext);
+    if (!windowsToConsider.isEmpty()) {
```

Comment:
It is a per-bundle cost of iterating in-memory O(<# of windows present in bundle, post-merge>) so it scales well with bundle size. Even if it scaled less well, I'd expect this to be unmeasurable relative to even to other overheads, much less nontrivial processing. FWIW my proposal changes the number of iterations from 3 to 4; I'm also OK with leaving it at 3 and merging `prefetchEmitPane` into `windowsThatShouldFire` which should elide any disagreement.

If we have measurements demonstrating this cost, then certainly we should document them and leave micro-optimizations in place. But given where the SDK is today, let's favor readability and abstraction over micro optimizations unless we have compelling evidence.